# Using timeit module 

### ns - nanoseconds $10^{-9} seconds$
### us - microseconds $10^{-6} seconds$
### ms - milliseconds $10^{-3} seconds$

#### Order of magnitutde(speed)

#### nanosecond(ns) > microsecond(us) > millisecond(ms)

### $10^{-9} > 10^-6 > 10 ^-3$

In [1]:
%timeit 2 + 3

29.2 ns ± 5.72 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


### Using Python List

In [2]:
%timeit [i**2 for i in range(1,4)]

2.71 µs ± 244 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


In [3]:
%timeit [i **2 for i in range(1,11)]

7.12 µs ± 1.62 µs per loop (mean ± std. dev. of 7 runs, 100000 loops each)


In [4]:
%timeit [i**2 for i in range(1,10000)]

6.91 ms ± 1.19 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)


### Using numpy arrays

In [5]:
import numpy as np

In [6]:
%timeit np.arange(1,4) ** 2

4.14 µs ± 873 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


In [7]:
%timeit np.arange(1,11) ** 2

4.11 µs ± 659 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


In [8]:
%timeit np.arange(1,10000) ** 2

20.6 µs ± 301 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)


We can clearly see that, 

In the first step, python list is faster when only there is a iteration over a list of only 3 numbers but 
as the loop increases with 10,000 numbers Numpy is faster than Python List.

nanosecond(ns) > microsecond(us) > millisecond(ms)

$10^{-9} > 10^-6 > 10 ^-3$

1) %timeit [i**2 for i in range(1,4)]   took 2.71 µs 
   
   %timeit np.arange(1,4) ** 2    took 4.14 µs
    
2) %timeit [i **2 for i in range(1,11)]   took 7.12 µs
   
   %timeit np.arange(1,11) ** 2   took 4.11 µs


3) %timeit [i**2 for i in range(1,10000)]   took 6.91 ms
   
   %timeit np.arange(1,10000) ** 2   took 20.6 µs   